### PR TITLE
client, daemon, overlord/devicestate: request system action API and stubs

### PR DESCRIPTION
**Pull request by bboozzoo**
The PR adds the API support for requesting system action. The devicemgr bit is mostly a stub that performs basic checks

### CODE REVIEW

File: client/systems.go [hunk 1]
```diff
@@ -70,3 +73,24 @@ func (client *Client) ListSystems() ([]System, error) {
 	}
 	return rsp.Systems, nil
 }
+
+// RequestSystemAction issues a request to run the given seed system in the
+// provided mode.
+func (client *Client) RequestSystemAction(systemLabel, actionMode string) error {
```
Comment (pedronis) [inline client/systems.go]: Maybe this should be something like:

DoSystemAction(sysAction *SystemAction) error ?

File: client/systems.go [hunk 2]
```diff
@@ -70,3 +73,24 @@ func (client *Client) ListSystems() ([]System, error) {
...
+	req := struct {
```
Comment (pedronis) [inline client/systems.go]: we want the option to add other possiblities to the post, also normally all the posts in our APIs tend to have a action field, so maybe somethind like:

```
   Action string  `json:"action"`
   *SystemAction
```

we would set Action="do"

File: client/systems.go (hunk 1 above)
Comment (bboozzoo) [reply client/systems.go]: As discussed on IRC we'll do `DoSystemAction(label, sysAction *SystemAction)` and POST to `/v2/systems/{label}`

Comment (bboozzoo) [reply client/systems.go]: This is done now.

File: overlord/devicestate/devicemgr.go [hunk 3]
```diff
@@ -780,29 +792,53 @@ func (m *DeviceManager) Systems() ([]System, error) {
...
+			sysAction = &act
+			break
+		}
+	}
+	if sysAction == nil {
+		return fmt.Errorf("cannot request unsupported mode %q for system %q",
+			action.Mode, systemLabel)
+	}
+
+	// TODO:UC20 request action
```
Comment (mvo5) [inline overlord/devicestate/devicemgr.go]: This comment could be expanded a little bit, I suppose this means that we actually set the right boot option and schedule a reboot?

File: daemon/api_systems.go [hunk 4]
```diff
@@ -82,3 +89,44 @@ func getSystems(c *Command, r *http.Request, user *auth.UserState) Response {
...
+
+	sa := devicestate.SystemAction{
+		Title: req.Title,
+		Mode:  req.Mode,
+	}
+	if err := c.d.overlord.DeviceManager().RequestSystemAction(systemLabel, sa); err != nil {
+		if os.IsNotExist(err) {
+			return NotFound("requested seed system %q does not exist", systemLabel)
+		}
+		// distinguish bad action?
```
Comment (mvo5) [inline daemon/api_systems.go]: If you edit this anyway, maybe adding a "XXX" ?

File: daemon/api_systems_test.go [hunk 5]
```diff
@@ -160,3 +172,97 @@ func (s *apiSuite) TestGetSystemsNone(c *check.C) {
...
+	d.overlord.AddManager(mgr)
+
+	s.vars = map[string]string{"label": "20191119"}
+	body := `{"action":"do","mode":"install"}`
+	req, err := http.NewRequest("POST", "/v2/systems/20191119", strings.NewReader(body))
+	c.Assert(err, check.IsNil)
+	rsp := postSystemsAction(systemsActionCmd, req, nil).(*resp)
+
+	c.Assert(rsp.Type, check.Equals, ResponseTypeError)
+	c.Check(rsp.Status, check.Equals, 404)
```
Comment (mvo5) [inline daemon/api_systems_test.go]: Could this test be moved into TestSystemActionRequestInvalid if that would grow a httpStatus field in `type table struct{ label, body, error string }` ? Or is there more to it? (it's fine if not, just curious as the test look very similar to the one that is already a table driven test).

File: daemon/api_systems_test.go [hunk 6]
```diff
@@ -160,3 +172,97 @@ func (s *apiSuite) TestGetSystemsNone(c *check.C) {
...
+
+	s.vars = map[string]string{"label": "20191119"}
+	body := `{"action":"do","title":"reinstall","mode":"install"}`
+	req, err := http.NewRequest("POST", "/v2/systems/20191119", strings.NewReader(body))
+	c.Assert(err, check.IsNil)
+	rsp := postSystemsAction(systemsActionCmd, req, nil).(*resp)
+	c.Check(rsp.Status, check.Equals, 200)
+}
+
+func (s *apiSuite) TestSystemActionBrokenSeed(c *check.C) {
```
Comment (mvo5) [inline daemon/api_systems_test.go]: Same question about integrating into table driven as TestSystemActionRequestNoSystem

File: overlord/devicestate/devicemgr.go (hunk 3 above)
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go]: yeah, I'll update the comment

File: daemon/api_systems.go (hunk 4 above)
Comment (bboozzoo) [reply daemon/api_systems.go]: actually I tweaked this to return a known error

File: daemon/api_systems_test.go (hunk 6 above)
Comment (bboozzoo) [reply daemon/api_systems_test.go]: Kept this one separate as the test intentionally breaks the seed.

File: daemon/api_systems_test.go (hunk 5 above)
Comment (bboozzoo) [reply daemon/api_systems_test.go]: Rolled this into a single test with other simple error scenarios.

File: client/systems.go [hunk 7]
```diff
@@ -70,3 +74,32 @@ func (client *Client) ListSystems() ([]System, error) {
...
+	}{
+		Action:       "do",
+		SystemAction: action,
+	}
+
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(&req); err != nil {
+		return err
+	}
+	if _, err := client.doSync("POST", "/v2/systems/"+systemLabel, nil, nil, &body, nil); err != nil {
```
Comment (stolowski) [inline client/systems.go]: Some wondering... Should we do some basic validation of systemLabel here? Out of curiosity I've checked existing client code and this lead to interesting errors ;), e.g.
```$ snap change ////
error: cannot unmarshal: json: cannot unmarshal array into Go value of type client.changeAndData
```

Comment (bboozzoo) [reply client/systems.go]: Hmmm, valid concern. I'll do a followup with checks in the `seed` package.

Comment (bboozzoo) [reply client/systems.go]: This is done in #8358

**Pull request merged**